### ISSUE TITLE
Suggested Lint: Excessive use of "pub(crate)"

### ISSUE THREAD

**Issue by kentfredric** [rust-lang/rust-clippy#3907]
I see lots of cases where an item is used within the crate, and a "pub(crate)" visibility specifier is used.

For current rust to not complain, all it requires is that the item is actually consumed somewhere (ie: not dead code).

But it also needs to possibly identify situations where the "pub(crate)" modifier is not needed due to all the call-sites being local, and thus the "pub(crate)" modifier can be stripped.

**Comment by flip1995** [comment]
This would also apply to things like `pub(super)` and so on, wouldn't it? 

So best case would be a lint `this doesn't need to be <visbility> and could be turned into <lesser-visiblity>`.

**Comment by Manishearth** [comment]
I'd be wary of adding such a lint, at least on by default, because internal APIs being used only in one file can be a temporary state that will change over time as the crate grows.

Overall a liberal sprinkling of pub(crate) isn't a bad thing, sometimes the modules are split up primarily for code organization reasons, not privacy reasons

**Comment by kentfredric** [comment]
Yeah, it wouldn't be a default lint, its more a lint designed for cleanup, specifically if your intention is to improve modularity by breaking reliance on pub(crate) where possible.

Sure, the alternative way to do this is just remove the declaration entirely and recompile and see where it breaks, but that's surely not going to be as an effective use of time.

**Comment by kentfredric** [comment]
> So best case would be a lint `this doesn't need to be <visbility> and could be turned into <lesser-visiblity>.`

Yes, that's more the gist of the idea.

**Comment by shepmaster** [comment]
I was about to open a similar issue, so I'll drop my example case here:

```rust

fn main() {
    inner::interface();
}

mod inner {
    pub fn interface() {
        only_used_in_inner();
    }
    
    pub(crate) fn only_used_in_inner() {}
}
```

In this code, the `pub(crate)` on `only_used_in_inner` allows the function to be used outside the `inner` module, but that capability is unused. I'd rather dial it down to be private inside the module.

A starting draft of message if anyone picks this up:

```
warning: function is more visible than necessary: `only_used_in_inner`
  --> src/main.rs:LL:CC
   |
LL |    pub(crate) fn only_used_in_inner() {
   |    ^^^^^^^^^^ help: remove this visibility
   |
   = note: `#[warn(excessively_visible))]` on by default
```

**Comment by joshtriplett** [comment]
I found myself needing this as well, to help me with cleanup.

Given that this would be crate-internal API, it seems like there's never value in having something marked as `pub(crate)` that isn't used outside the module; an internal API could always be changed from private to `pub(crate)` or vice versa without impacting external API surface or semver.

**Comment by zjp-CN** [comment]
`clippy::redundant_pub_crate` can already catch this pattern, so this issue has been fixed?

```rust
#![deny(clippy::redundant_pub_crate)]

fn main() {
    inner::interface();
}

mod inner {
    pub fn interface() {
        only_used_in_inner();
    }
    
    pub(crate) fn only_used_in_inner() {}
}
```

```rust
error: pub(crate) function inside private module
  --> src/main.rs:12:5
   |
12 |     pub(crate) fn only_used_in_inner() {}
   |     ----------^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     help: consider using: `pub`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.97.0/index.html#redundant_pub_crate
note: the lint level is defined here
  --> src/main.rs:1:9
   |
 1 | #![deny(clippy::redundant_pub_crate)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
